### DISCUSSION TITLE
How to use a coupledArrayGradient ?

### DISCUSSION THREAD

**Discussion by ngrilli** [idaholab/moose#20686] [Q&A General] [answered]
Dear MOOSE Users,

I am trying to implement an AuxKernel that calculates the directional derivative of an array variable.
For this purpose, I am using the coupledArrayGradient that I saw in one example in the MOOSE framework postprocessors.

However, there is no example where such a coupled gradient is actually used,
not just imported, in the code in the entire MOOSE,
therefore I have doubt about how to use it, specifically, how to access the individual components of the gradient.

Let's say I have:

`_grad_variable(coupledArrayGradient("gradient_variable"))`

I assume the first index should the quadrature point, so:

`_grad_variable[_qp]`

and the second index should be in round parenthesis to pick the gradient of a specific variable in the vector array, like:

`_grad_variable[_qp](i)`

Is this correct? Or i means the 3D component, like x,y,z, along which I am taking the gradient?

How do I access the individual components?

You can check my class here:
https://github.com/ngrilli/c_pfor_am/blob/main/src/auxkernels/ArrayDirectionalDerivative.C

Right now I assign the expression `_grad_variable[_qp](i)` to a RealVectorValue
and extract the components with round parenthesis,
I would like to know what is the best strategy in your opinion and if what I am doing is correct.

Specifically, since I need the directional derivative, I would need the scalar product between `_grad_variable[_qp](i)`
and a `std::vector<Real>`. What is the best strategy for this.

Your suggestions are highly appreciated.
Hope this can be helpful for others as I could not find any discussion here
with keyword `coupledArrayGradient`.
Thank you very much in advance.

Best Regards,
Nicolò

**Comment by GiudGiud** [comment]
Hello

Here's an example of using it
https://github.com/idaholab/moose/blob/c42b80999f9a1a7aaa09c96484fc8a760f992359/framework/src/postprocessors/ElementIntegralArrayVariablePostprocessor.C

so `[_qp](_component)` for _u. Lemme check for grad_u

Guillaume

**Reply by GiudGiud** [reply]
So for grad_u it should be:

`_grad_u[_qp](_gradient_component * <NUM_ARRAY_COMPONENTS> + _component)`

If you do 
`RealVectorValue bb = _grad_u[_qp](1)`
you'll get a vector filled with (`_grad_u[_qp](1)`, 0 , 0)

**Reply by GiudGiud** [reply]
So this vector<Real> is indexed over spatial dimension I think based on what I see in your code. A simple for-loop like you did sounds fine for this operation, though I think there is an issue with your cast to RealVectorValue of `_grad_u[_qp](i)`.
And the for loop needs to skip indexes in the grad_vector to stay on the same component.

**Reply by GiudGiud** [reply]
Hold that thought I ll double check this 
EDIT: was not what I thought. Previous messages edited so the answer's right

**Reply by ngrilli** [reply]
@GiudGiud  thank you for your answers.
Tomorrow we will go through some systematic testing and check that the gradient is calculated as expected.
For the moment I can only tell that this code works and when I print _grad_u [_qp] (i)
I am always getting a 3D vector, but as you point out, the behaviour may not be what I expect
and I may need to use (_gradient_component * <NUM_ARRAY_COMPONENTS> + _component)
instead.
I will post again tomorrow once we do the testing
Nicolò

**Comment by ngrilli** [comment, marked as answer]
I am answering this as I figured out.

Referring to the variable _grad_variable(coupledArrayGradient("gradient_variable"))

If gradient_variable is an ArrayVariable with n component,
then _grad_variable is an array variable with 3n components.

The order of the components of _grad_variable is the following:
1) the first n components are the derivatives along x of the components of gradient_variable
2) the next n components are the derivatives along y of the components of gradient_variable
3) the next n components are the derivatives along z of the components of gradient_variable

For instance a directional derivatives of an ArrayVariable are calculated as shown here:
https://github.com/ngrilli/c_pfor_am/blob/main/src/auxkernels/ArrayDirectionalDerivative.C

Nicolò
